### PR TITLE
Move pkg/backend/openshiftclusters/create into cluster installation

### PR DESCRIPTION
**Pull request by hawkowl**
### Which issue this PR addresses:

Part of the backend refactoring work https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/6118707

### What this PR does / why we need it:

Moves pkg/backend/openshiftclusters/create into a step that's run by pkg/clusters/install , and removes the pkg/backend/openshiftclusters abstraction that's now replaced by pkg/cluster.

### Test plan for issue:

Hopefully E2E tests cover it. Also adding new tests.

### Is there any documentation that needs to be updated for this PR?

N/A

### CODE REVIEW

File: pkg/cluster/generateconfig.go [hunk 1]
```diff
@@ -28,143 +24,67 @@ import (
...
 	"github.com/Azure/ARO-RP/pkg/util/subnet"
 	"github.com/Azure/ARO-RP/pkg/util/version"
 )
 
-func (m *manager) Create(ctx context.Context) error {
-	var err error
+func (m *manager) checkQuotas(ctx context.Context) error {
+	// we don't re-call Dynamic on subsequent entries here.  One reason is
+	// that we would re-check quota *after* we had deployed our VMs, and
+	// could fail with a false positive.
```
Comment (jim-minter) [inline pkg/cluster/generateconfig.go:39]: I think the PR breaks this.  If the RP crashes after deploy starts and the process restarts on another backend, quota will be checked again and may fail the second time around.

File: pkg/cluster/generateconfig.go [hunk 2]
```diff
@@ -28,143 +24,67 @@ import (
...
+	ocDynamicValidator := validate.NewOpenShiftClusterDynamicValidator(m.fpAuthorizer, m.log, m.env, m.doc.OpenShiftCluster, m.subscriptionDoc)
+	return ocDynamicValidator.Dynamic(ctx)
```
Comment (jim-minter) [inline pkg/cluster/generateconfig.go:41]: hmm, I can see this has to be instantiated here, but it's ugly.  Should think about this as a follow-up.

File: pkg/cluster/install.go [hunk 3]
```diff
@@ -46,13 +45,32 @@ func (m *manager) AdminUpgrade(ctx context.Context) error {
...
+		installConfig *installconfig.InstallConfig
+		platformCreds *installconfig.PlatformCreds
+		image         *releaseimage.Image
+	)
+
 	steps := map[api.InstallPhase][]steps.Step{
 		api.InstallPhaseBootstrap: {
+			steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.checkQuotas)),
+			steps.Action(m.generateSSHKey),
+			steps.Action(func(ctx context.Context) error {
```
Comment (jim-minter) [inline pkg/cluster/install.go:66]: missing `steps.Action(m.ensureACRToken),` ?

File: pkg/cluster/generateconfig.go [hunk 4]
```diff
@@ -28,143 +24,67 @@ import (
...
-		return nil
-	})
-	if err != nil {
-		return err
-	}
-
 	pullSecret, err := pullsecret.Build(m.doc.OpenShiftCluster, string(m.doc.OpenShiftCluster.Properties.ClusterProfile.PullSecret))
 	if err != nil {
-		return err
+		return nil, nil, nil, err
```
Comment (jim-minter) [inline pkg/cluster/generateconfig.go:50]: ugly as hell but I'm +1 on it as an interim step.

**Review by jim-minter — CHANGES_REQUESTED** [review]
I can only see a couple of issues with this, nice work.  However, this was a bit daunting to review (although I know I dove in while it was still in Draft).  Completely separate commits that did one transformation at a time and were completely self-contained would have been really welcome.  The absence of them + the types of issues I spotted makes me a little anxious about merging this at the end of this sprint; I wonder if we should carry this over to after the next RP release has been done?  WDYT?

File: pkg/cluster/generateconfig.go (hunk 2 above)
Comment (jim-minter) [reply pkg/cluster/generateconfig.go:41]: (don't attempt to solve this on this PR - too much change in one go)

File: pkg/cluster/generateconfig.go (hunk 1 above)
Comment (hawkowl) [reply pkg/cluster/generateconfig.go]: I think I've fixed this.

File: pkg/cluster/install.go (hunk 3 above)
Comment (hawkowl) [reply pkg/cluster/install.go:65]: fix'd

File: pkg/cluster/generateconfig.go (hunk 4 above)
Comment (hawkowl) [reply pkg/cluster/generateconfig.go:55]: we can get rid of at least 1 of these with 4.6, at least :)

File: pkg/cluster/generateconfig.go (hunk 2 above)
Comment (hawkowl) [reply pkg/cluster/generateconfig.go:47]: yeah :(

File: pkg/cluster/install.go [hunk 5]
```diff
@@ -44,13 +43,33 @@ func (m *manager) AdminUpgrade(ctx context.Context) error {
 	return m.runSteps(ctx, steps)
 }
 
+func (m *manager) Update(ctx context.Context) error {
+	// no-op
+	return nil
+}
+
```
Comment (mjudeikis) [inline pkg/cluster/install.go:50]: Rebase errror. It is missing changed from https://github.com/Azure/ARO-RP/pull/1083/files
```
func (m *manager) Update(ctx context.Context) error {
	var err error
	timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Minute)
	defer cancel()
	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
		err = m.ocDynamicValidator.Dynamic(ctx)
		if azureerrors.HasAuthorizationFailedError(err) ||
			azureerrors.HasLinkedAuthorizationFailedError(err) {
			m.log.Print(err)
			return false, nil
		}
		return err == nil, err
	}, timeoutCtx.Done())
	if err != nil {
		return err
	}

	return nil
}
```

File: pkg/api/validate/openshiftcluster_validatedynamic.go [hunk 6]
```diff
@@ -39,12 +39,7 @@ type OpenShiftClusterDynamicValidator interface {
 }
 
 // NewOpenShiftClusterDynamicValidator creates a new OpenShiftClusterDynamicValidator
-func NewOpenShiftClusterDynamicValidator(log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster, subscriptionDoc *api.SubscriptionDocument) (OpenShiftClusterDynamicValidator, error) {
-	fpAuthorizer, err := env.FPAuthorizer(subscriptionDoc.Subscription.Properties.TenantID, env.Environment().ResourceManagerEndpoint)
-	if err != nil {
-		return nil, err
-	}
-
+func NewOpenShiftClusterDynamicValidator(fpAuthorizer refreshable.Authorizer, log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster, subscriptionDoc *api.SubscriptionDocument) OpenShiftClusterDynamicValidator {
```
Comment (mjudeikis) [inline pkg/api/validate/openshiftcluster_validatedynamic.go:42]: As you will be changing can you move `fpAuthorizer` to be after `log` and `env`. At some point, we tried to keep order like: `ctx`, `log`,`env`, `other stuff` for all signatures.

File: pkg/cluster/generateconfig.go [hunk 7]
```diff
@@ -28,143 +24,72 @@ import (
...
-				return err
-			}
-		}
+func (m *manager) checkQuotas(ctx context.Context) error {
+	// we don't re-call Dynamic on subsequent entries here.  One reason is that
+	// we would re-check quota *after* we had deployed our VMs, and could fail
+	// with a false positive. We detect if this is the first time by only
+	// running if the cluster doesn't have a set SSH key (which will be
+	// configured in a later step).
+	if m.doc.OpenShiftCluster.Properties.SSHKey != nil {
```
Comment (jim-minter) [inline pkg/cluster/generateconfig.go:42]: previously this was `if m.doc.OpenShiftCluster.Properties.Install == nil {` -- what's the rationale for the change?

Comment (jim-minter) [reply pkg/cluster/generateconfig.go:42]: nm, I think I understand it

File: pkg/cluster/generateconfig.go [hunk 8]
```diff
@@ -28,143 +24,72 @@ import (
...
-
-			m.doc, err = m.db.PatchWithLease(ctx, m.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
-				m.acrtoken.PutRegistryProfile(doc.OpenShiftCluster, rp)
-				return nil
-			})
-			if err != nil {
-				return err
-			}
-		}
+func (m *manager) checkQuotas(ctx context.Context) error {
```
Comment (jim-minter) [inline pkg/cluster/generateconfig.go:36]:
```suggestion
func (m *manager) dynamicValidate(ctx context.Context) error {
```
?

Comment (jim-minter) [reply pkg/cluster/generateconfig.go:36]: Olga's change (which we think is mis-rebased here) added dynamic validation to Update as well.  Given that dynamic validation is now called on Create and Update, does this architecture where dynamic validate is called as part of the pkg/cluster sequence still make sense?

This PR would be moving to:
Create -> func (m *manager) Install -> runSteps -> checkQuotas()
Update -> call dynamic validator

Which is a slightly odd difference.  Would it make more sense for Update -> func (m *manager) Update -> runSteps -> checkQuotas() as well?

In this case, we might want an explicit switch in checkQuotas where we do all the dynamic validation and only exclude the quota step if SSHKey != nil ?

We probably want to get to the point where we have table-based steps in Update, and where the dynamic validation is shared between Install and Update -- I guess I'm saying, should that be done now?

**Comment by jim-minter** [comment]
@hawkowl this is close, if you can get the changes in soon we'll try and get this merged in the next day or two.

**Comment by hawkowl** [comment]
@jim-minter @mjudeikis I've refactored the quota checking and dynamic validator a little bit. It feels like the quota checking and the dynamic validator as we use them now are actually a bit orthogonal, so I've tried splitting them out into something that makes sense with how we want to use it going forward.

File: pkg/cluster/acrtoken.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	pkgacrtoken "github.com/Azure/ARO-RP/pkg/util/acrtoken"
+	"github.com/Azure/ARO-RP/pkg/util/deployment"
+)
+
+func (m *manager) ensureACRToken(ctx context.Context) error {
+	if m.env.DeploymentMode() == deployment.Development {
+		return nil
+	}
+
+	acrtoken, err := pkgacrtoken.NewManager(m.env, m.localFpAuthorizer)
```
Comment (jim-minter) [inline pkg/cluster/acrtoken.go:19]: follow-up: acrtoken should be a field of *manager, instantiated in cluster.go

File: pkg/api/validate/quota.go [hunk 10]
```diff
@@ -78,24 +83,59 @@ func addRequiredResources(requiredResources map[string]int, vmSize api.VMSize, c
 	return nil
 }
 
-// validateQuotas checks usage quotas vs. resources required by cluster before cluster creation
-func (dv *openShiftClusterDynamicValidator) validateQuotas(ctx context.Context) error {
-	dv.log.Print("validateQuotas")
+type AzureQuotaValidator interface {
```
Comment (jim-minter) [inline pkg/api/validate/quota.go:86]: nit: tempted just to call this Validator

File: pkg/api/validate/quota.go [hunk 11]
```diff
@@ -78,24 +83,59 @@ func addRequiredResources(requiredResources map[string]int, vmSize api.VMSize, c
...
+
+type quotaValidator struct {
+	log *logrus.Entry
+	env env.Interface
+
+	oc      *api.OpenShiftCluster
+	spUsage compute.UsageClient
+}
+
+func NewAzureQuotaValidator(ctx context.Context, log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster) (AzureQuotaValidator, error) {
```
Comment (jim-minter) [inline pkg/api/validate/quota.go:98]: nit: NewQuotaValidator

File: pkg/cluster/install.go [hunk 12]
```diff
@@ -44,13 +43,37 @@ func (m *manager) AdminUpgrade(ctx context.Context) error {
 	return m.runSteps(ctx, steps)
 }
 
+func (m *manager) Update(ctx context.Context) error {
```
Comment (jim-minter) [inline pkg/cluster/install.go:46]: follow-up: there probably needs to be a rebalancing of functions across cluster.go, install.go and some new files.

**Review by jim-minter — APPROVED** [review]
Terrific.  Merging.
